sermant-plugins/sermant-mq-grayscale/sermant-mq-grayscale-plugin/src/main/java/io/sermant/mq/grayscale/utils/SubscriptionDataUtils.java

```diff
@@ -0,0 +1,418 @@
+        Map<String, List<String>> trafficTagMap = new HashMap<>();
+        for (Map.Entry<String, String> entry : item.getTrafficTag().entrySet()) {
+            buildTrafficTagMap(trafficTagMap, entry);
+        }
+        StringBuilder builder = new StringBuilder();
+        if (trafficTagMap.size() > 1) {
+            builder.append(LEFT_BRACKET);
+        }
+        for (Map.Entry<String, List<String>> envEntry : trafficTagMap.entrySet()) {
+            if (builder.length() > 1) {
+                builder.append(" or ");
+            }
+            builder.append(LEFT_BRACKET)
+                    .append(envEntry.getKey())
+                    .append(" in ")
+                    .append(getStrForSets(new HashSet<>(envEntry.getValue())))
+                    .append(RIGHT_BRACKET);
+        }
+        if (trafficTagMap.size() > 1) {
+            builder.append(RIGHT_BRACKET);
+        }
+        return builder.toString();
+    }
+
+    private static String buildBaseConsumerSql92Expression(List<GrayTagItem> items) {
+        Map<String, List<String>> trafficTagMap = new HashMap<>();
+        for (GrayTagItem item : items) {
+            for (Map.Entry<String, String> entry : item.getTrafficTag().entrySet()) {
+                buildTrafficTagMap(trafficTagMap, entry);
+            }
+        }
+        StringBuilder builder = new StringBuilder();
+        if (trafficTagMap.size() > 1) {
+            builder.append(LEFT_BRACKET);
+        }
+        for (Map.Entry<String, List<String>> envEntry : trafficTagMap.entrySet()) {
+            if (builder.length() > 1) {
+                builder.append(AND_SPLICE_STR);
+            }
+            if (trafficTagMap.size() > 1) {
+                builder.append(LEFT_BRACKET);
+            }
+            builder.append(LEFT_BRACKET)
+                    .append(envEntry.getKey())
+                    .append(" not in ")
+                    .append(getStrForSets(new HashSet<>(envEntry.getValue())))
+                    .append(RIGHT_BRACKET)
+                    .append(" or ")
+                    .append(LEFT_BRACKET)
+                    .append(envEntry.getKey())
+                    .append(" is null")
+                    .append(RIGHT_BRACKET);
+            if (trafficTagMap.size() > 1) {
+                builder.append(RIGHT_BRACKET);
+            }
+        }
+        if (trafficTagMap.size() > 1) {
+            builder.append(RIGHT_BRACKET);
+        }
+        return builder.toString();
+    }
+
+    private static void buildTrafficTagMap(Map<String, List<String>> trafficTagMap, Map.Entry<String, String> entry) {
+        if (trafficTagMap.containsKey(entry.getKey())) {
+            trafficTagMap.get(entry.getKey()).add(entry.getValue());
+        } else {
+            List<String> list = new ArrayList<>();
+            list.add(entry.getValue());
+            trafficTagMap.put(entry.getKey(), list);
+        }
+    }
```

Comment:
just write like this: trafficTagMap.computeIfAbsent(entry.getKey(), k -> new ArrayList<>()).add(entry.getValue());

Reply:
fixed